File: C/Linked List/Doubly Linked List/DoublyLL03.c
/*
=====================================================================

Name            : Doubly Linked List - Two Nodes Linking
Program Number  : 003
Description     : This program demonstrates creation of two nodes
                  in a Doubly Linked List and establishes links
                  between them using next and prev pointers.

                  The first node points to the second node through
                  the next pointer, while the second node points
                  back to the first node through the prev pointer.

Concepts Used   : Doubly Linked List, Self-Referential Structure,
                  Dynamic Memory Allocation, Pointer Linking,
                  Forward Link, Backward Link

Input           : None

Output          : Displays data of both nodes using links

Date            : July 2026
Language        : C
Author          : Atharv Tushar Bhosale

=====================================================================
*/

#include<stdio.h>
#include<stdlib.h>

#pragma pack(1)

struct node
{
    int data;
    struct node *next;
    struct node *prev;
};

typedef struct node NODE;
typedef struct node* PNODE;

int main()
{
    PNODE first = NULL;
    PNODE second = NULL;

    first = (PNODE)malloc(sizeof(NODE));
    second = (PNODE)malloc(sizeof(NODE));

    first->data = 11;
    first->next = second;
    first->prev = NULL;

    second->data = 21;
    second->next = NULL;
    second->prev = first;

    printf("First Node Data  : %d\n", first->data);
    printf("Second Node Data : %d\n", first->next->data);

    free(first);
    free(second);

    return 0;
}

/*
=====================================================================

Sample Output :

First Node Data  : 11
Second Node Data : 21

=====================================================================

Node Representation :

NULL <- |11| <=> |21| -> NULL

=====================================================================

Note :

- first->next stores address of second node.
- second->prev stores address of first node.
- This creates a bidirectional connection between nodes.
- Doubly Linked List supports traversal in both directions.

=====================================================================
*/

File: Java/Linked List/Doubly Linked List/DoublyLL03.java
/*
===============================================================================
Program Name    : Doubly Linear Linked List - Insert Last
Program Number  : 003
File Name       : DoublyLL03.java

Description     : This program demonstrates the implementation of the
                  InsertLast() operation in a Doubly Linear Linked List
                  using Java. The new node is inserted at the end of
                  the linked list while maintaining both next and
                  previous links.

Date            : July 2026
Language        : Java
Author          : Atharv Tushar Bhosale
===============================================================================

Concepts Used:
1. Class and Object
2. Self-Referential Class
3. Dynamic Memory Allocation
4. Doubly Linear Linked List
5. Insert Last Operation
6. Traversing Linked List
7. Pointer Manipulation
8. Encapsulation

Key Learning Points:
- Create a new node dynamically.
- Traverse to the last node.
- Insert the new node at the end.
- Update next and previous links.
- Display the linked list.
- Count total number of nodes.

===============================================================================
*/

class node
{
    public int data;
    public node next;
    public node prev;

    node(int no)
    {
        this.data = no;
        this.next = null;
        this.prev = null;
    }
}

class DoublyLL
{
    private node first;
    private int iCount;

    public DoublyLL()
    {
        first = null;
        iCount = 0;
    }

    public void Display()
    {
    }

    public int Count()
    {
        return iCount;
    }

    public void InsertLast(int iNo)
    {
        node newn = new node(iNo);
        node temp = first;

        if(first == null)
        {
            first = newn;
        }
        else
        {
            while(temp.next != null)
            {
                temp = temp.next;
            }

            temp.next = newn;
            newn.prev = temp;
        }

        iCount++;
    }
}

class DoublyLL03
{
    public static void main(String A[])
    {
        DoublyLL dobj = new DoublyLL();

        dobj.InsertLast(11);
        dobj.InsertLast(21);
        dobj.InsertLast(51);
        dobj.InsertLast(101);

        dobj.Display();

        System.out.println("Number of Nodes : " + dobj.Count());
    }
}

/*
===============================================================================
Expected Output:
===============================================================================

| 11 | <=> | 21 | <=> | 51 | <=> | 101 | <=> NULL
Number of Nodes : 4

===============================================================================
Memory Representation:
===============================================================================

first
 │
 ▼

NULL <- [11] <=> [21] <=> [51] <=> [101] -> NULL

===============================================================================
Algorithm:
===============================================================================

1. Create a new node.
2. If the list is empty, make it the first node.
3. Otherwise, traverse till the last node.
4. Link the new node after the last node.
5. Update the previous pointer.
6. Increment node count.
7. Display the linked list.

===============================================================================
Time Complexity:
===============================================================================

InsertLast()  : O(n)

Display()     : O(n)

Count()       : O(1)

Space Complexity : O(1)

===============================================================================
Conclusion:
===============================================================================

Program 003 demonstrates the InsertLast() operation in a Doubly Linear
Linked List using Java. It appends a new node at the end of the list
while correctly maintaining both next and previous links.

===============================================================================
*/
